src/gradualizer_lib.erl

```diff
@@ -100,11 +102,13 @@ reverse_graph(G) ->
 %% `UserTy' is actually an unexported `gradualizer_type:af_user_defined_type()'.
 
 -spec get_type_definition(UserTy, Env, Opts) -> {ok, Ty} | opaque | not_found when
-      UserTy :: gradualizer_type:abstract_type(),
+      UserTy :: type(),
       Env :: typechecker:env(),
       Opts :: [annotate_user_types],
-      Ty :: gradualizer_type:abstract_type().
+      Ty :: type().
 get_type_definition({remote_type, _Anno, [{atom, _, Module}, {atom, _, Name}, Args]}, _Env, _Opts) ->
+    %% We matched out the non-type() elements above so we can assert Args :: [type()]
+    Args = ?assert_type(Args, [type()]),
```

Comment:
Why? Atom literals are part of abstract_type().

Reply:
Atom literals can be represented by `type()`, but literal atoms are not of type `type()`, see https://github.com/josefs/Gradualizer/blob/3cda2e5aa1b1313b3e02fce67986fcda3ab9a02c/src/gradualizer_type.erl#L290-L292

Reply:
Ahh, sorry, the above is bollocks, as the atoms here are actually in the AST representation. Let me check again...

Reply:
Ok, so we start with this error:

```
$ gradualizer -I include/ -- src/gradualizer_lib.erl
src/gradualizer_lib.erl: The variable on line 112 at column 43 is expected to have type [type()]
but it has type [abstract_type()] | {atom, anno(), atom()}

    %% We matched out the non-type() elements above so we can assert Args :: [type()]
    %Args = ?assert_type(Args, [type()]),
    gradualizer_db:get_type(Module, Name, Args);
                                          ^^^^
```

We check the type of a remote call node:
https://github.com/josefs/Gradualizer/blob/3cda2e5aa1b1313b3e02fce67986fcda3ab9a02c/src/gradualizer_type.erl#L254-L257

So `Args` is a `[type()]`, not a `type()`, which is fine for passing down. Singleton atoms are not fine, as we need a list. We know we've already matched on the singleton atoms, so by elimination we're safe to assert `Args :: [type()]`.

Reply:
Great! Only the comment needs an update then. E.g. `%% We matched out the single atom arguments, so only Args (list-of-type) remains`.